update(driver): add support for chown syscall family

yo348 opened the pull request:
**What type of PR is this?**

/kind feature

**Any specific area of the project related to this PR?**

/area driver-kmod
/area driver-bpf
/area driver-modern-bpf
/area libpman
/area libsinsp

**Does this PR require a change in the driver versions?**

/version driver-API-version-minor
/version driver-SCHEMA-version-minor

**What this PR does / why we need it**:

Adds full processing for `chown`, `lchown`, `fchown`, `fchownat` syscalls extracting their parameters: filename, new user id and group id.

**Which issue(s) this PR fixes**:

Fixes #892

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
new: add support for chown, lchown, fchown, fchownat syscalls
```

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/905#" title="Author self-approved">yo348</a>*
**Once this PR has been reviewed and has the lgtm label**, please assign jasondellaluce for approval. For more information see [the Kubernetes Code Review Process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process).

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["jasondellaluce"]} -->

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [6265c40](https://github.com/falcosecurity/libs/commits/6265c407f90b996c4f7e4c45d1e7526549a7b491) fix test: libsinsp event_table

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

yo348 commented:
Not sure why (not modern) bpf test for fchownat_x fails on arm

yo348 commented on driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/chown.bpf.c:
```diff
@@ -0,0 +1,80 @@
+
+	/* Parameter 2: filename (type: PT_FSPATH) */
+	unsigned long path_pointer = extract__syscall_argument(regs, 0);
+	auxmap__store_charbuf_param(auxmap, path_pointer, MAX_PATH, USER);
```
I created these by analogy with chmod.bpf.c that also use auxmap. My understanding is that anything that has a file path is variable length (because path can be short like `/bin/sh` or long like `/opt/one/two/three/asdf/bash`). And the function `auxmap__store_charbuf_param` reads a variable length string up to `MAX_PATH`

https://github.com/falcosecurity/libs/blob/c592f4f230e48f36a50c1716fd94e7e279b67513/driver/modern_bpf/helpers/store/auxmap_store_params.h#L361

Andreagit97 replied:
yep here we should use `auxmap` approach :+1:

Andreagit97 commented:
thank you for this, i will take a look ASAP :)

Andreagit97 commented on driver/bpf/fillers.h:
```diff
@@ -5778,6 +5786,107 @@ FILLER(sys_fchmod_x, true)
+	CHECK_RES(res);
+
+	/* Parameter 4: gid (type: PT_UINT32) */
+	u32 gid = (u32)bpf_syscall_get_argument(data, 3);
```
here you miss a `res = bpf_val_to_ring(data, gid);` this is the reason why tests are failing on arm64

yo348 replied:
Thanks!

yo348 commented on test/drivers/test_suites/syscall_exit_suite/fchownat_x.cpp:
```diff
@@ -0,0 +1,59 @@
+	const char* pathname = "*//null";
+	uint32_t uid = 0;
+	uint32_t gid = 0;
+	uint32_t flags = 0;
```
Good idea.
`fcntl.h` is already included via `event_class.h`

yo348 commented on driver/ppm_flag_helpers.h:
```diff
@@ -1711,6 +1711,16 @@ static __always_inline u32 chmod_mode_to_scap(unsigned long modes)
+{
+	u32 res = 0;
+
+	if (flags & AT_SYMLINK_FOLLOW)
```
Didn't realize at first that `AT_EMPTY_PATH` is supported, but I see in linux source `do_fchmodat` checks for it. Thanks

yo348 commented on driver/API_VERSION:
```diff
@@ -1 +1 @@
-4.0.0
+4.1.0
```
removed this

yo348 commented on driver/ppm_fillers.c:
```diff
@@ -6749,6 +6749,160 @@ int f_sys_fchmod_x(struct event_filler_arguments *args)
+	if (unlikely(res == PPM_FAILURE_INVALID_USER_MEMORY))
+	{
+		res = val_to_ring(args, (unsigned long)"<NA>", 0, false, 0);
+	}
```
removed

yo348 commented on userspace/libscap/examples/01-open/scap_open.c:
```diff
@@ -145,6 +145,10 @@ static int simple_set[] = {
+	PPM_SC_CHOWN,
+	PPM_SC_LCHOWN,
+	PPM_SC_FCHOWN,
+	PPM_SC_FCHOWNAT,
```
removed this, it is not a necessary part for this functionality anyway and the simple set can be expanded at any time later

yo348 commented on userspace/libsinsp/events/sinsp_events_ppm_sc.cpp:
```diff
@@ -119,6 +119,10 @@ libsinsp::events::set<ppm_sc_code> libsinsp::events::enforce_simple_sc_set(libsi
 		PPM_SC_UNSHARE,
 		PPM_SC_USERFAULTFD,
 		PPM_SC_VFORK,
+		PPM_SC_CHOWN,
```
removed this, it is not a necessary part for this functionality anyway and the simple set can be expanded at any time later

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/905#" title="Author self-approved">yo348</a>*


The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

<details open>
Needs approval from an approver in each of these files:


Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/905#" title="Author self-approved">yo348</a>*
**Once this PR has been reviewed and has the lgtm label**, please assign jasondellaluce for approval. For more information see [the Kubernetes Code Review Process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process).

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["jasondellaluce"]} -->

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/905#" title="Author self-approved">yo348</a>*
**Once this PR has been reviewed and has the lgtm label**, please assign jasondellaluce for approval. For more information see [the Kubernetes Code Review Process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process).

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["jasondellaluce"]} -->

poiana commented:
Adding label `do-not-merge/contains-merge-commits` because PR contains merge commits, which are not allowed in this repository.
Use `git rebase` to reapply your commits on top of the target branch. Detailed instructions for doing so can be found [here](https://git.k8s.io/community/contributors/guide/github-workflow.md#4-keep-your-branch-in-sync).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/905#" title="Author self-approved">yo348</a>*


The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

<details open>
Needs approval from an approver in each of these files:


Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/905#" title="Author self-approved">yo348</a>*
**Once this PR has been reviewed and has the lgtm label**, please assign jasondellaluce for approval. For more information see [the Kubernetes Code Review Process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process).

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["jasondellaluce"]} -->

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/905#" title="Author self-approved">yo348</a>*
**Once this PR has been reviewed and has the lgtm label**, please assign jasondellaluce for approval. For more information see [the Kubernetes Code Review Process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process).

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["jasondellaluce"]} -->

yo348 commented:
@Andreagit97 thank you for the quick and thorough review, and finding the issues I missed. I just rebased on top of master (hence the mess with PR closing and reopening) and addressed your concerns. Please take a look when you have time.

Andreagit97 commented on driver/ppm_flag_helpers.h:
```diff
@@ -1711,6 +1711,21 @@ static __always_inline u32 chmod_mode_to_scap(unsigned long modes)
+{
+	u32 res = 0;
+
+	if (flags & AT_SYMLINK_FOLLOW)
+		res |= PPM_AT_SYMLINK_FOLLOW;
```
```suggestion
#ifdef AT_SYMLINK_FOLLOW
	if (flags & AT_SYMLINK_FOLLOW)
		res |= PPM_AT_SYMLINK_FOLLOW;
#endif
```

Andreagit97 reviewed (commented):
Left a real minor comment :)

Andreagit97 commented:
Feel free to create a new fresh commit instead of rewriting the history otherwise we cannot see what is changed :/

Andreagit97 reviewed (approved):
/approve
thank you!

poiana commented:
LGTM label has been added.  <details>Git tree hash: 811cc6e91fb36ef30c9e5fb893c4c5a7b71dbb2b</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/905#pullrequestreview-1306315827" title="Approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/libs/pull/905#" title="Author self-approved">yo348</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Andreagit97]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

FedeDP commented on driver/bpf/fillers.h:
```diff
@@ -5776,6 +5784,108 @@ FILLER(sys_fchmod_x, true)
+	CHECK_RES(res);
+
+	/* Parameter 2: path (type: PT_FSPATH) */
+	unsigned long path_pointer = bpf_syscall_get_argument(data, 0);
```
Don't we need to resolve the user ptr and take the path here?

FedeDP replied:
And the others path pointers from `lchown`, `fchownat`

FedeDP replied:
Wow but they are tested :laughing: am i missing anything?

Andreagit97 replied:
we resolve them in the `bpf_val_to_ring` :)

FedeDP replied:
Whoa i completely forgot about that! Then it LGTM!

FedeDP reviewed (approved):
/approve

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/905#pullrequestreview-1306315827" title="Approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/libs/pull/905#pullrequestreview-1307023500" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/libs/pull/905#" title="Author self-approved">yo348</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Andreagit97,FedeDP]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
